xtrace-example/src/connection_inner.rs

```diff
@@ -0,0 +1,279 @@
+            _ => eprintln!("Unknown server setup response: {:?}", packet),
+        }
+    }
+
+    /// Handle a request sent by the client
+    pub fn client_request(&mut self, packet: &[u8]) {
+        fn do_parse(inner: &mut ConnectionInner, packet: &[u8]) -> Result<(), ParseError> {
+            let seqno = inner.next_client_request;
+            inner.next_client_request = seqno.wrapping_add(1);
+
+            let (header, remaining) = parse_request_header(packet, BigRequests::Enabled)?;
+            let request = Request::parse(header, remaining, &mut Vec::new(), &inner.ext_info)?;
+            println!("client ({}): {:?}", seqno, request);
+
+            // Is this a QueryExtension?
+            let queried_extension = if let Request::QueryExtension(ref request) = request {
+                match String::from_utf8(request.name.to_vec()) {
+                    Ok(name) => {
+                        println!("Extension name: {}", name);
+                        Some(name)
+                    }
+                    Err(e) => {
+                        println!("Extension name is not utf8: {:?}", e);
+                        None
+                    }
+                }
+            } else {
+                None
+            };
+
+            // Does the request have a reply? If so, remember it.
+            if let Some(parser) = request.reply_parser() {
+                inner.pending_replies.push_back(PendingReply {
+                    seqno,
+                    parser,
+                    queried_extension,
+                });
+            }
+
+            Ok(())
+        }
+        if let Err(e) = do_parse(self, packet) {
+            eprintln!("Error while parsing a client request: {:?}", e);
+        }
+    }
+
+    /// Handle an X11 error sent by the server
+    pub fn server_error(&mut self, packet: &[u8]) {
+        fn do_parse(inner: &mut ConnectionInner, packet: &[u8]) -> Result<(), ParseError> {
+            let err = Error::parse(packet, &inner.ext_info)?;
+            println!("server ({}): {:?}", err.wire_sequence_number(), err);
+
+            // Remove a pending request if it failed
+            let next_pending = inner.pending_replies.front().map(|r| r.seqno);
+            if next_pending == Some(err.wire_sequence_number()) {
+                let _ = inner.pending_replies.pop_front();
+            }
+
+            Ok(())
+        }
+        if let Err(e) = do_parse(self, packet) {
+            eprintln!("Error while parsing an X11 error: {:?}", e);
+        }
+    }
+
+    /// Handle an X11 event sent by the server
+    pub fn server_event(&mut self, packet: &[u8]) {
+        fn do_parse(inner: &mut ConnectionInner, packet: &[u8]) -> Result<(), ParseError> {
+            let event = Event::parse(packet, &inner.ext_info)?;
+            println!(
+                "server ({}): {:?}",
+                event.wire_sequence_number().unwrap_or(0),
+                event,
+            );
```

Comment:
Would it be worth to use `{:#?}` to have pretty printing?

Reply:
Dunno. I didn't even know this `#` option exists.
With `{:?}`:
```
client (324): FreeGC(FreeGCRequest { gc: 31457297 })
client (325): FreeGC(FreeGCRequest { gc: 31457295 })
client (326): RenderFreePicture(FreePictureRequest { picture: 31457312 })
client (327): DestroyWindow(DestroyWindowRequest { window: 31457289 })
client (328): FreePixmap(FreePixmapRequest { pixmap: 31457308 })
client (329): RenderFreePicture(FreePictureRequest { picture: 31457309 })
client (330): FreeCursor(FreeCursorRequest { cursor: 31457282 })
client (331): FreeGC(FreeGCRequest { gc: 31457280 })
client (332): CloseFont(CloseFontRequest { font: 31457283 })
client (333): GetInputFocus(GetInputFocusRequest)
server (327): UnmapNotify(UnmapNotifyEvent { response_type: 18, sequence: 327, event: 31457289, window: 31457289, from_configure: false })
```
With `{:#?}`:
```
client (369): GetInputFocus(
    GetInputFocusRequest,
)
server (363): UnmapNotify(
    UnmapNotifyEvent {
        response_type: 18,
        sequence: 363,
        event: 35651593,
        window: 35651593,
        from_configure: false,
    },
)
server (363): FocusOut(
    FocusInEvent {
        response_type: 10,
        detail: Ancestor,
        sequence: 363,
        event: 35651593,
        mode: Normal,
    },
)
server (363): DestroyNotify(
    DestroyNotifyEvent {
        response_type: 17,
        sequence: 363,
        event: 35651593,
        window: 35651593,
    },
)
server (369): GetInputFocus(
    GetInputFocusReply {
        revert_to: None,
        sequence: 369,
        length: 0,
        focus: 8388877,
    },
)
```
You don't want to see the connection setup in this alternative mode. `XkbGetMapReply` is also quite long.

I guess both versions are hard to read, but the `{:?}`-version is at least simpler to scroll through to find something.

Reply:
Ideally, there should be command line flag to select the format.

Anyways, as an example I think it is good enough for now and the format can be changed in a future PR if needed.